cmd/kubeadm/app/phases/controlplane/users_linux.go

```diff
@@ -0,0 +1,260 @@
+// +build linux
+
+package controlplane
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"strconv"
+	"strings"
+
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/utils/exec"
+)
+
+var ()
+
+var UpdateFileOwnershipFn = func(filepath string, uid int64, gid int64, permissions uint32) error {
+	if err := os.Chown(filepath, int(uid), int(gid)); err != nil {
+		return err
+	}
+	if err := os.Chmod(filepath, fs.FileMode(permissions)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func UpdateFileOwnership(filepath string, uid int64, gid int64, permissions uint32) error {
+	return UpdateFileOwnershipFn(filepath, uid, gid, permissions)
+}
+
+var UpdateDirecotryOwnershipFn = func(directorypath string, uid int64, gid int64, permissions string) error {
+	exec := exec.New()
+	cmd := exec.Command("chown", "-R", fmt.Sprintf("%d:%d", uid, gid), directorypath)
+	buff := bytes.Buffer{}
+	cmd.SetStderr(&buff)
+	if err := cmd.Run(); err != nil {
+		return errors.New("failed to change directory owner: " + err.Error() + " : " + buff.String())
+	}
+	cmd = exec.Command("chmod", "-R", permissions, directorypath)
+	buff = bytes.Buffer{}
+	cmd.SetStderr(&buff)
+	if err := cmd.Run(); err != nil {
+		return errors.New("failed to change directory permission: " + err.Error() + " : " + buff.String())
+	}
+	return nil
+}
+
+func UpdateDirecotryOwnership(directorypath string, uid int64, gid int64, permissions string) error {
+	return UpdateDirecotryOwnershipFn(directorypath, uid, gid, permissions)
+}
+
+func getControlPlaneUserId(username string) (int64, bool, error) {
```

Comment:
all the functions in this file would need unit test coverage.
instead of hardcoding "/etc/passwd" it could be a constant on the side of callers such as GetUserAndGroupIDs

and having a signature like:
```
func getControlPlaneUserId(passwdFile, username string) (int64, bool, error) {
```

then you could unit test getControlPlaneUserId with a file from a temporary directory.

Reply:
...but why not use https://golang.org/pkg/os/user/ for group / user look up?
the package is fairly stable on Linux.